### PR TITLE
 python@3.12: promote to default python3

### PR DESCRIPTION
**Pull request by branchv**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

---

Hopefully this is a smoother transition than last year since all formulae should now be using fully-qualified python binaries like `python3.11`

### CODE REVIEW

**Comment by branchvincent** [comment]
So far without Linux
* bpytop: known https://github.com/Homebrew/homebrew-core/pull/121636#issuecomment-1407815158, will look into fixing this 
* condure: seen before, will look into fixing this
* pyqt@5: #150515
* trino: known, https://github.com/trinodb/trino/issues/18983983

**Comment by Bo98** [comment]
Might be worth revision bumping so the user doesn't end up in a scenario where they have Python 3.12 already installed but no `python3`.

File: Formula/p/python@3.12.rb
Comment (Bo98) [inline Formula/p/python@3.12.rb:481]: Does this work anymore with the externally managed thing?

**Comment by onlined** [comment]
What is blocking this PR right now? I can lend a hand.

**Comment by nickmccurdy** [comment]
This appears to still be good.

**Comment by nickmccurdy** [comment]
This appears to still be good. I'm still not sure why we wouldn't want to migrate to the newest version, especially considering the typing improvements.

**Review by SMillerDev — CHANGES_REQUESTED** [review]
Awaiting https://github.com/Homebrew/homebrew-core/pull/150390#discussion_r1354994844

**Comment by cho-m** [comment]
> Awaiting [#150390 (comment)](https://github.com/Homebrew/homebrew-core/pull/150390#discussion_r1354994844)

Which may also depend on https://github.com/Homebrew/homebrew-core/issues/157500 if we decide to drop majority of `python-*` formulae.

File: Formula/p/python@3.12.rb
Comment (iMichka) [reply Formula/p/python@3.12.rb:481]: Doc update is being worked on here: https://github.com/Homebrew/brew/pull/16643

Comment (iMichka) [reply Formula/p/python@3.12.rb]: Doc update is done. I went for a caveats cleanup: if people use pip incorrectly there will be a message about virtualenvs and the externally managed process IMHO.

**Comment by iMichka** [comment]
Re-running CI to get a clearer view of what is still broken

**Comment by cho-m** [comment]
> Re-running CI to get a clearer view of what is still broken

On macOS, `gtk-doc` is failing:
```
  ==> Testing gtk-doc
  ==> /usr/local/Cellar/gtk-doc/1.33.2_1/bin/gtkdoc-scan --module=test
  Failed to execute: /usr/local/Cellar/gtk-doc/1.33.2_1/bin/gtkdoc-scan
```

Probably needs a shebang fix (preferably using a stable path across migrations):
```sh
#!@@HOMEBREW_PREFIX@@/opt/python@3.12/libexec/bin/python3
```

**Comment by alex** [comment]
This was not actually fixed.

**Comment by branchvincent** [comment]
Oops, didn't mean to close it

**Comment by branchvincent** [comment]
Previous run: https://github.com/Homebrew/homebrew-core/actions/runs/7952028325?pr=150390

On linux:

* `gtk-doc`: Fixed via #163303
* `libcap-ng`: `No module named 'capng'` (unrelated; seen previously, thought I fixed with #162479 but seems python bindings aren't being installed in the bottle)
* `llvm@13`: `No module named 'clang'`

We still need `revision` bumps and to fix `llvm@13`, but we should be good to merge after one more run

**Comment by p-linnane** [comment]
I've made sure we have an available long timeout label if you want to push this through now @branchvincent.

**Review by iMichka — APPROVED** [review]

**Review by SMillerDev — APPROVED** [review]

File: Formula/p/python@3.12.rb
Comment (cho-m) [inline Formula/p/python@3.12.rb:475]: Did we want to remove this caveat? There is also the `python-gdbm` caveat in `python@3.11` that didn't carry over to `python@3.12`.

`python-tk@3.12` and `python-gdbm@3.12` will remain as they are part of base Python but we just put them into separate formulae (given that Homebrew doesn't support sub/split-packages).

Not important to handle here but may want to add caveats in follow up.

Comment (p-linnane) [reply Formula/p/python@3.12.rb:475]: I agree that it makes sense to handle this in a followup syntax PR.

**Review by p-linnane — APPROVED** [review]

File: Formula/p/python@3.12.rb
Comment (iMichka) [reply Formula/p/python@3.12.rb:475]: I removed that line because it now felt useless. We had it for a few years now I think. I hope our users have adjusted their workflows, and my hope was to stop telling obvious things in the caveats.
But if you think we need to add it back: this can be done in a syntax-only follow-up pull request.

Comment (cho-m) [reply Formula/p/python@3.12.rb:475]: It doesn't matter to me whether we keep or remove it.

Homebrew's way of separate formulae is a bit odd since we don't support sub/split-packages. Most repositories would either install everything together (e.g. Arch Linux) or use subpackages (which colocates the related packages that use the same source distribution).

**Comment by cho-m** [comment]
Failures that can be followed up on:
- `libcap-ng` is probably installing Python bindings into `$HOMEBREW_PREFIX/lib/...` rather than `#{lib}/...` since we don't have a sandbox on Linux. This means it passes when built from source but will fail later on when installed from bottle.
- `profanity` is building with `python3-embed` so Python feature is disabled on macOS and linked to `python@3.11` on Linux. May be worth adding a major-versioned symlink to `libexec` so we can use `PKG_CONFIG_PATH` rather than an `inreplace` workaround (or manually passing in corresponding `python_CFLAGS`/`python_LIBS`).

**Comment by p-linnane** [comment]
Thanks for all the work here @branchvincent!

**Comment by branchvincent** [comment]
And thanks @iMichka for all the doc updates!

**Pull request merged**
